### PR TITLE
chore(release): v2.60.0

### LINKED ISSUE #292: [Documentation] Missing: How to remove node properties using updateNode (undefined pattern)
Hej, wrzucam kolejne zgłoszenie po tym jak ai sobie nie radził i robiliśmy śledztwo. Najpierw pełne zgłoszenie napisane przez Claude po całej sprawie, potem podrzucę historię.
// Zbyszek

### 🏷️ Title:
```
[Documentation] Missing: How to remove node properties using updateNode (undefined pattern)
```

---

### 📝 Description:

```markdown
## Problem Description

The documentation for `n8n_update_partial_workflow` with `updateNode` operations doesn't explain how to **remove** properties from nodes. This leads to confusion when trying to migrate from deprecated properties (e.g., `continueOnFail` → `onError`).

## What We Tried

**Scenario**: Update HTTP Request node from deprecated `continueOnFail: true` to modern `onError: "continueRegularOutput"`

**Attempt 1** - Only set the new property:
```javascript
{
  type: "updateNode",
  updates: {
    onError: "continueRegularOutput"
  },
  nodeName: "getUnmanagedGroups"
}
```

**Expected**: Old `continueOnFail` would be removed or ignored  
**Actual**: Both properties remained on the node, causing validation error:
```
"Cannot use both continueOnFail and onError properties. Use only onError for modern workflows."
```

**Attempt 2** - Explicitly set deprecated property to false:
```javascript
{
  type: "updateNode",
  updates: {
    onError: "continueRegularOutput",
    continueOnFail: false
  },
  nodeName: "getUnmanagedGroups"
}
```

**Expected**: Setting to `false` would disable the old behavior  
**Actual**: Same validation error - validator checks for **presence** of both properties, not their values

## Our Investigation

We used DeepWiki to analyze the n8n-mcp source code and found:

### 1. The Solution Exists in Code
File: `src/services/workflow-auto-fixer.ts` (L311-317)
```typescript
const operation: UpdateNodeOperation = {
  type: 'updateNode',
  nodeId: nodeName,
  updates: {
    onError: undefined // This will remove the property
  }
};
```

**The pattern exists but is undocumented for users.**

### 2. Validation Logic
File: `src/services/workflow-validator.ts` (L1385-1393)
```typescript
if (node.continueOnFail !== undefined && node.onError !== undefined) {
    result.errors.push({
      message: 'Cannot use both "continueOnFail" and "onError" properties.'
    });
}
```

Validator checks for property **existence**, not values. Both `true` and `false` trigger the error.

### 3. Why This Isn't Intuitive
Most REST APIs use "update = replace" semantics. Users naturally assume:
- Not mentioning a property = remove it
- Setting to `false` = disable it

But n8n-mcp uses **selective merge** - only provided properties are touched, others remain unchanged.

### Add Documentation Section: "Removing Node Properties"

Location: `tools_documentation` for `n8n_update_partial_workflow`, after "Update node parameter" example

**Suggested content**:

```markdown
### Removing Properties

To remove a property from a node, explicitly set it to `undefined`:

**Example 1: Migrate from deprecated continueOnFail to modern onError**
```javascript
n8n_update_partial_workflow({
  id: "workflow-id",
  operations: [{
    type: "updateNode",
    nodeName: "HTTP Request",
    updates: {
      onError: "continueRegularOutput",
      continueOnFail: undefined  // Removes deprecated property
    }
  }]
})
```

**Example 2: Remove error handling completely**
```javascript
n8n_update_partial_workflow({
  id: "workflow-id",
  operations: [{
    type: "updateNode",
    nodeName: "MyNode",
    updates: {
      onError: undefined,
      retryOnFail: undefined,
      maxTries: undefined
    }
  }]
})
```

**Note**: The `updateNode` operation uses selective merge - it only updates properties you explicitly provide. Properties not mentioned remain unchanged. Use `undefined` to explicitly remove properties.
```

### Additional Context

**Why this matters**:
1. Common migration scenario (n8n deprecated `continueOnFail` in favor of `onError`)
2. Validator enforces mutual exclusivity but docs don't explain removal pattern
3. Pattern is used internally (auto-fixer) but not exposed to users
4. Natural assumptions (update = replace) lead to frustration

**Related documentation**:
- Changelog mentions error handling update (2025-07-09) but doesn't explain removal pattern
- Node-specific validator warns about deprecated properties but doesn't show how to remove them

## Environment
- n8n-mcp version: 2.18.1
- n8n instance: https://n8n.bitback.cloud
- Node.js: v22.15.0

## Workaround
Once we discovered the `undefined` pattern through source code analysis, it worked perfectly:

```javascript
{
  type: "updateNode",
  updates: {
    onError: "continueRegularOutput",
    continueOnFail: undefined  // ✅ This removes the property
  },
  nodeName: "getUnmanagedGroups"
}
```

The workflow validated successfully after this change.
```

---

**Gotowe do wklejenia.** Wszystko co trzeba w jednym issue - problem, śledztwo, rozwiązanie, sugestia poprawki.

### LINKED ISSUE #714: Per-operation tool filtering for read-only deployments (DISABLED_TOOL_OPERATIONS)
## Background

Several users running n8n-mcp in enterprise / governance-sensitive contexts want a true read-only mode but are blocked by tools that bundle read and write operations under a single tool name. The existing `DISABLED_TOOLS` env var only operates at the tool level, so disabling a destructive branch means losing the read branch as well.

Discussed in #564 (by @mahmoudnaif, with a strong governance writeup from @christiansousadev).

## The problem

Two tools mix read and destructive operations under one name:

- **`n8n_workflow_versions`**
  - Read: `list`, `get`
  - Write/destructive: `rollback`, `delete`, `prune`, `truncate`

- **`n8n_executions`**
  - Read: `list`, `get`
  - Destructive: `delete`

Today the only way to block the destructive branches via n8n-mcp is to disable the entire tool, which also kills the valuable read paths. Tool annotations (`destructiveHint: true`) exist but are metadata only — nothing in the dispatch layer enforces them.

The robust workaround today is n8n-side API key scoping (RBAC), which is the right pattern for production but does not give clients a way to express the same intent at the MCP layer.

## Proposed feature

Add an env var that filters operations *within* a tool at dispatch time, before the handler runs:

```
DISABLED_TOOL_OPERATIONS=n8n_workflow_versions:delete,rollback,prune,truncate;n8n_executions:delete
```

Format sketch (open to bikeshedding):
- Semicolon-separated list of `<tool_name>:<comma_separated_operations>`
- Operation names match the existing `action` / `mode` enum values in each tools schema
- Disabled operations should also be advertised as such in `tools_documentation` output so the LLM does not waste a turn trying them

## Acceptance criteria

- [ ] New env var parsed at startup with the same safety limits as `DISABLED_TOOLS` (size cap, count cap, validation against known tool/operation names)
- [ ] Dispatch layer in `src/mcp/server.ts` checks the disabled-operation set before calling the handler and returns a clear error (something like `Operation 'delete' on tool 'n8n_executions' is disabled by server policy`)
- [ ] `tools_documentation` reflects the disabled operations so they are visible to clients
- [ ] Unit tests covering: parser edge cases, dispatch enforcement for both `n8n_workflow_versions` and `n8n_executions`, interaction with existing `DISABLED_TOOLS`
- [ ] README / `docs/` section explaining the read-only deployment recipe (this env var + n8n API key RBAC)

## Out of scope

- Full RBAC inside n8n-mcp — that lives in the n8n API
- Per-user / per-session policy — this is a server-level config knob

## Notes

- The implementation is straightforward: a Set lookup in the dispatch path. Cheap at runtime.
- Happy to review a PR. Tagging @christiansousadev and @mahmoudnaif from the discussion in case either of you wants to take a swing at it.

Concieved by Romuald Członkowski — https://www.aiadvisors.pl/en

### PR DESCRIPTION
**Pull request by czlonkowski**
## Release v2.60.0

Version bump + CHANGELOG entry for everything merged into `main` since **v2.59.4**. Minor bump because this release adds a new feature (`DISABLED_TOOL_OPERATIONS`).

> ⚠️ Merging this PR triggers `release.yml` (it watches `package.json` / `package.runtime.json` on `main`), which publishes to npm and GHCR. Merge only when you're ready to ship.

### What's in this release

**Added**
- Per-operation tool filtering via `DISABLED_TOOL_OPERATIONS` — disable individual operations within a tool (e.g. `n8n_executions:delete`) for read-only / least-privilege deployments (#719, closes #714) — thanks @mahmoudnaif

**Fixed**
- Diff engine now removes a property when a patch sets it to `undefined` (#794, closes #292) — thanks @justadityaraj
- Version-summary cache TTL was 1000× too long (passed ms to a seconds API) (#804) — thanks @linda-ai-bot
- Google Sheets validator accepts the `columns` resourceMapper (#730) — thanks @infobewaize
- Code node return validator no longer fooled by helper functions / methods / regex / comments / `for await` (#795) — thanks @AjTheSpidey

**Changed**
- Hardened Dependabot config: exclude n8n packages, add `/ui-apps` entry, `rebase-strategy: auto` (#874)
- Dev/CI tooling bumps: `ui-apps` vite 8 / typescript 6 / plugin-react 6 (#876–#878, verified end-to-end), GitHub Actions (#869–#873)

**Documentation**
- Multi-client `npx` npm cache contention note in `docs/SELF_HOSTING.md` (#866)

### Verification
- `npm run build` ✅ clean
- Full unit suite ✅ **5023 passed** / 35 skipped, coverage 81.18%
- Version consistent across `package.json`, `package-lock.json`, `package.runtime.json`, `manifest.json` (regenerated, 24 tools)
- `ui-apps` combined upgrade build verified in an isolated worktree (all 5 apps)

Conceived by Romuald Członkowski - https://www.aiadvisors.pl/en

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,30 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- **Version-summary cache TTL was 1000× too long** (#804). The node version-summary cache passed its TTL to `SimpleCache.set(key, data, ttlSeconds)` in milliseconds (`86400000`), but the API treats the third argument as seconds and multiplies by 1000 internally — so the entry was scheduled to live ~1,000 days instead of the intended 24 hours, and stale version summaries would never expire within a normal process lifetime. The TTL is now passed in seconds (`86400`). Thanks to @linda-ai-bot for the fix.
+- **Google Sheets validator accepts the `columns` resourceMapper** (#730). `validateGoogleSheetsUpdate` raised false-positive `missing_required` errors on valid Google Sheets v4+ configurations that map data with the `columns` resourceMapper (`mappingMode: "defineBelow"` / `"autoMapInputData"`) instead of the legacy `range` + `values` fields. The validator now recognises the `columns` mapping shape for `update`/`append`, while `read` continues to require only `range`. Thanks to @infobewaize for the fix.
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply CHANGELOG.md]: Good catch — corrected in 62e2de0. The actual new behavior in #730 is on `update` (it now treats a real `columns` mapping as satisfying both `range` and `values`). `append` already accepted `columns` and was only tightened so an empty `columns: {}` no longer passes; `read` is unchanged and still requires `range`. The entry now reflects that precisely.
